sdk/common/src/main/java/io/opentelemetry/sdk/internal/AttributeUtil.java

```diff
@@ -93,6 +129,48 @@ public static Object applyAttributeLengthLimit(Object value, int lengthLimit) {
       String str = (String) value;
       return str.length() < lengthLimit ? value : str.substring(0, lengthLimit);
     }
+    if (value instanceof Value) {
+      return applyValueLengthLimit((Value<?>) value, lengthLimit);
+    }
+    return value;
+  }
+
+  @SuppressWarnings("unchecked")
+  private static Value<?> applyValueLengthLimit(Value<?> value, int lengthLimit) {
+    ValueType type = value.getType();
+
+    if (type == ValueType.STRING) {
+      String str = (String) value.getValue();
+      if (str.length() <= lengthLimit) {
+        return value;
+      }
+      return Value.of(str.substring(0, lengthLimit));
+    } else if (type == ValueType.BYTES) {
+      ByteBuffer buffer = (ByteBuffer) value.getValue();
+      int length = buffer.remaining();
+      if (length <= lengthLimit) {
+        return value;
+      }
+      byte[] truncated = new byte[lengthLimit];
+      buffer.get(truncated);
+      return Value.of(truncated);
+    } else if (type == ValueType.ARRAY) {
+      List<Value<?>> array = (List<Value<?>>) value.getValue();
```

Comment:
The `AttributeUtil.applyAttributesLimit` function takes an approach of first checking if any any attribute is invalid before paying the price to create a new `AttributeBuilder` and rebuilding, with the limits applied to each entry. 

Could take the same approach here to avoid unnecessary reconstruction of the `Value` if all the entries are valid. Applies to `KEY_VALUE_LIST` as well.

Don't necessarily need to solve it now, but let's at least leave a TODO.

Reply:
https://github.com/open-telemetry/opentelemetry-java/pull/7814/commits/d44e727421cbc9073afa697183ab5c12dd8837c5